src/object-store/src/util.rs

```diff
@@ -78,7 +78,55 @@ pub fn normalize_dir(v: &str) -> String {
 /// - Otherwise, it's a file path.
 pub fn join_path(parent: &str, child: &str) -> String {
     let output = format!("{parent}/{child}");
-    opendal::raw::normalize_path(&output)
+    normalize_path(&output)
+}
+
+/// Modified from the `opendal::raw::normalize_path`
+///
+/// # The different
+///
+/// It only keeps `/` ahead if the original path starts with `/`.
+///
+/// Make sure all operation are constructed by normalized path:
+///
+/// - Path endswith `/` means it's a dir path.
+/// - Otherwise, it's a file path.
+///
+/// # Normalize Rules
+///
+/// - All whitespace will be trimmed: ` abc/def ` => `abc/def`
+/// - **(Removed❗️)** ~~ All leading / will be trimmed: `///abc` => `abc`~~
+/// - Internal // will be replaced by /: `abc///def` => `abc/def`
+/// - Empty path will be `/`: `` => `/`
+pub fn normalize_path(path: &str) -> String {
+    // - all whitespace has been trimmed.
+    let path = path.trim();
+
+    // Fast line for empty path.
+    if path.is_empty() {
+        return "/".to_string();
+    }
+
+    let has_leading = path.starts_with('/');
+    let has_trailing = path.ends_with('/');
+
+    let mut p = path
+        .split('/')
+        .filter(|v| !v.is_empty())
+        .collect::<Vec<&str>>()
+        .join("/");
+
+    // If path is not starting with `\` but it should
+    if !p.starts_with('/') && has_leading {
+        p.insert(0, '/');
+    }
+
+    // If path is not ending with `\` but it should
+    if !p.ends_with('/') && has_trailing {
+        p.push('/');
+    }
```

Comment:
```suggestion
    if has_leading {
        p.insert(0, '/');
    }

    if has_trailing {
        p.push('/');
    }
```

After `filter not is_empty` and `join`, p is always in form `a/b/c`.

Reply:
Making those changes breaks the following unit test: 
```
assert_eq!("/", join_path("", ""));

assertion `left == right` failed
  left: "/"
 right: "//"
```

Because both `has_leading` and `has_trailing` will be set to true for a single `/`, and them both will be inserted again at the end, resulting at `//`, I believe this is not the expected behavior, could you confirm?

Reply:
Thanks for sharing this edge case. You are right that your code is better in such case.

While I'm thinking we may check the input since `join_path("", "")` doesn't seem a valid input for me, you code can handle all the case.

cc @v0y4g3r @evenyag I wonder if we should return error on `join_path("", "")` or assert it won't happen, or return `"/"`.

Reply:
Returning `/` is fine.